pkg/attestation/crafter/runners/oidc/github.go

```diff
@@ -0,0 +1,201 @@
+			SkipClientIDCheck: true,
+		}), nil
+	}
+	return &c, nil
+}
+
+// Token requests an OIDC token from GitHub's provider, verifies it, and returns the token.
+func (c *GitHubOIDCClient) Token(ctx context.Context) (*Token, error) {
+	tokenBytes, err := c.requestToken(ctx, defaultAudience)
+	if err != nil {
+		return nil, err
+	}
+
+	tokenPayload, err := c.decodePayload(tokenBytes)
+	if err != nil {
+		return nil, err
+	}
+
+	t, err := c.verifyToken(ctx, defaultAudience, tokenPayload)
+	if err != nil {
+		return nil, err
+	}
+
+	token, err := c.decodeToken(t)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := c.verifyClaims(token); err != nil {
+		return nil, err
+	}
+
+	token.RawToken = tokenPayload
+
+	return token, nil
+}
+
+func (c *GitHubOIDCClient) newRequestURL(audience []string) string {
+	requestURL := *c.requestURL
+	q := requestURL.Query()
+	for _, a := range audience {
+		q.Add("audience", a)
+	}
+	requestURL.RawQuery = q.Encode()
+	return requestURL.String()
+}
+
+func (c *GitHubOIDCClient) requestToken(ctx context.Context, audience []string) ([]byte, error) {
+	req, err := http.NewRequest("GET", c.newRequestURL(audience), nil)
+	if err != nil {
+		return nil, fmt.Errorf("%w: creating request: %w", errRequestError, err)
+	}
+	req.Header.Add("Authorization", "bearer "+c.bearerToken)
+	req = req.WithContext(ctx)
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return nil, fmt.Errorf("%w: %w", errRequestError, err)
+	}
+	defer resp.Body.Close()
+
+	b, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, fmt.Errorf("%w: reading response: %w", errRequestError, err)
+	}
+	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
+		return nil, fmt.Errorf("%w: response: %s: %s", errRequestError, resp.Status, string(b))
+	}
+	return b, nil
+}
+
+func (c *GitHubOIDCClient) decodePayload(b []byte) (string, error) {
+	var payload struct {
+		Value string `json:"value"`
+	}
+	decoder := json.NewDecoder(bytes.NewReader(b))
+	if err := decoder.Decode(&payload); err != nil {
+		return "", fmt.Errorf("%w: parsing JSON: %w", errToken, err)
+	}
+	return payload.Value, nil
+}
+
+// verifyToken verifies the token contents and signature.
+func (c *GitHubOIDCClient) verifyToken(ctx context.Context, audience []string, rawIDToken string) (*oidc.IDToken, error) {
+	// Verify the token.
+	verifier, err := c.verifierFunc(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("%w: creating verifier: %w", errVerify, err)
+	}
+
+	idToken, err := verifier.Verify(ctx, rawIDToken)
+	if err != nil {
+		return nil, fmt.Errorf("verify: could not verify token: %w", err)
+	}
+
+	// Verify the audience received is the one we requested.
+	if !compareStringSlice(audience, idToken.Audience) {
+		return nil, fmt.Errorf("%w: audience not equal %q != %q", errVerify, audience, idToken.Audience)
+	}
+
+	return idToken, nil
+}
+
+func (c *GitHubOIDCClient) decodeToken(token *oidc.IDToken) (*Token, error) {
+	var t Token
+	t.Issuer = token.Issuer
```

Comment:
Do you need to do explicit assignment here? I believe this could be done by embedding the registered claims, otherwise you might be dropping some information other than issuer, audience and expiry?

https://github.com/chainloop-dev/chainloop/blob/6660e0bf45e3ece3619dd08929097125ddc774a4/internal/robotaccount/cas/robotaccount.go#L35

Reply:
Refactored.